[DEV-17228] Add markup and content for premium plugins promos

michalkleiner opened the pull request:
### Description:

Follow up PR for #21401

- Add content and preview images for premium plugins
- Align widget names and image names with plugin names (with two exceptions for heatmaps and session recordings)
- Add template for plugin promo widget
- Extend ContentBlock component to allow for image with a 50/50 grid (this is needed so that the whole card can use the grid and the image is aligned at the top with the heading)
- Integrate 'Learn more' link opening a modal with plugin details
- Integrate 'Start free trial' CTA opening a new tab with plugin info from the marketplace
- Add UI tests

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

michalkleiner committed:
Add initial template for plugin promos

michalkleiner committed:
Move image to Card so grid can be used at card level

michalkleiner committed:
Add plugin promo screenshots

michalkleiner committed:
Adjust content

michalkleiner committed:
Use widget template instead of controller view

michalkleiner committed:
Populate plugin promo content

michalkleiner committed:
Remove enable/disable variable override

michalkleiner committed:
Remove legal info markup as not yet designed

michalkleiner committed:
Unify image alt text and remove help url property

michalkleiner committed:
Prepare promo widgets for refactoring for reusability

michalkleiner committed:
Integrate CTA links

michalkleiner committed:
Rename A/B Testing image to match plugin name

michalkleiner committed:
Use marketplace plugin info where possible

michalkleiner committed:
Update Start free trial CTA to open marketplace in new tab

michalkleiner committed:
Move promo features text to lang file

michalkleiner committed:
Add UI tests

michalkleiner commented on plugins/ProfessionalServices/Widgets/PromoAbTesting.php:
```diff
@@ -9,27 +9,42 @@
     public function render()
     {
-        return 'content';
+        $marketplacePlugins = StaticContainer::get('Piwik\Plugins\Marketplace\Plugins');
+        $pluginInfo = $marketplacePlugins->getPluginInfo(self::PROMO_PLUGIN_NAME);
```
TL;DR: This ultimately extends the time to see the promo as it does a marketplace API call.

I think it's the right way to go about getting any plugin related info like the link to its detail page on plugins.matomo.org or its display name, however, it slows it down.

We can hardcode the information here for the 2-3 fields we use in the template for each plugin and display the ad/promo content faster. The risk is that information might become out of sync with the marketplace around the plugin display name, the URL on which the user can purchase purchase it etc.

What are the opinions on this, @Stan-vw @caddoo @bx80?

caddoo replied:
I would say make it static for now, if we then have a problem with syncing (how often do those values really change?) then we go for the slower approach.
Its many sequential calls to the marketplace so I can imagine the delay is a lot

Stan-vw replied:
How are the current links in the marketplace coded, e.g. are they hardcoded or are they pulled from the API?
I suggest we stay consistent, rather than reinventing the wheel.

bx80 replied:
Another problem with static data would be the delay with people deploying new versions on-premise, so if static plugin data ever did change then an invalid name/link could be shown for months or years before it can be fixed by a new version.

 Also worth noting that [plugin data should be cached](https://github.com/matomo-org/matomo/blob/2c878f46ec3e0d270e4ae4231ff31721dacc7dd9/plugins/Marketplace/Api/Client.php#L266) after the first request so the delay should be much less for subsequent requests.

michalkleiner replied:
Seems like we're on the fence with this so I'd leave it as is (it's safer) and we can always hardcode it later if we need to.

michalkleiner committed:
Update UI test screenshots from CI

michalkleiner commented:
Also there might be some potential to abstract some parts of the widget configure and render methods into a base promo widget class of sorts. I looked into that and the complicating factor is that the configure method is static and render is instance, so if we e.g. wanted to provide new methods for the promoted plugin name or its display name etc, it would again be a mix of static and instance methods that can't be called from a parent class. Perhaps I didn't see the right way through it but it felt it was becoming too convoluted for little benefit.

michalkleiner commented:
@bx80 do you have any high level remarks/comments on this piece? Keen to get this in. Ta!

michalkleiner commented on plugins/ProfessionalServices/Widgets/PromoAbTesting.php:
```diff
@@ -9,27 +9,42 @@
     {
         $config->setCategoryId('ProfessionalServices_PromoAbTesting');
         $config->setSubcategoryId('ProfessionalServices_PromoOverview');
-        $config->setName('ProfessionalServices_PromoAbTestingOverview');
```
Removing the name here (and all other widgets) as that is only shown when the widget is loading, so for ease of future refactoring it's ok not to have it.

With it, the loading says "Loading data..." and then it changes to "Loading A/B Testing...", then the widget shows.
Without it, it shows "Loading data..." and then the widget shows.

I think it's even a better experience if it doesn't flicker with the plugin name, so I removed it.

mneudert commented on plugins/ProfessionalServices/tests/UI/ProfessionalServices_PluginPromo_spec.js:
```diff
@@ -0,0 +1,93 @@
+
+        await page.goto(urlBase + 'category=' + category + '&subcategory=' + subcategory);
+        await page.waitForNetworkIdle();
+        expect(await page.screenshotSelector('.pageWrap, .pluginPromo')).to.matchImage('promo_abtesting');
```
Can we limit the screenshots to just `.pluginPromo`?

The `.pageWrap` seems to not add much value but much grey space, probably matching the height of the navigation.

michalkleiner replied:
Yeah, I can remove that, it's a copy from some other UI test. The images are likely taller than they need to be because of that.

michalkleiner committed:
Simplify UI test selector

michalkleiner committed:
Update UI tests

michalkleiner commented on config/environment/test.php:
```diff
@@ -49,10 +49,9 @@
-
         $dontUseTestConfig = $c->get('test.vars.dontUseTestConfig');
         if (!$dontUseTestConfig) {
+            $testingEnvironment = $c->get('Piwik\Tests\Framework\TestingEnvironmentVariables');
```
This can be reset, it was done during a change that was reverted, but no harm in keeping this.

michalkleiner commented on core/Widget/WidgetsList.php:
```diff
@@ -139,7 +139,15 @@ public function addToContainerWidget($containerId, WidgetConfig $widget)
     {
         foreach ($this->widgets as $index => $widget) {
-            if ($widget->getCategoryId() === $widgetCategoryId) {
+            if (
+                ($widget->getCategoryId() === $widgetCategoryId)
+                ||
+                (
+                    (substr_compare($widgetCategoryId, '*', -1) === 0)
+                    &&
+                    (substr_compare($widget->getCategoryId(), $widgetCategoryId, 0, strlen($widgetCategoryId) - 1) === 0)
+                )
+            ) {
```
I can revert this and remove all the promo widgets one by one, or keep this and add a test for this and update the docblock.

bx80 commented on plugins/ProfessionalServices/lang/en.json:
```diff
@@ -4,12 +4,40 @@
+        "CTALearnMore": "Learn more about %1$s",
+        "AbTestingFeature01": "Stop guessing, start using data, and compare webpage versions to increase conversions.",
+        "AbTestingFeature02": "Compare various elements, including layout, design, copy, and more, while monitoring performance in real-time.",
+        "AbTestingFeature03": "Identify and address user behaviour patterns, so you can increase conversion rates and achieve better results from your digital efforts.",
+        "CrashAnalyticsFeature02": "Automatically track website crashes (e.g. broken carts, unresponsive forms, etc.) for swift bug resolution, so you can ensure a seamless and bug-free user experience.",
+        "CrashAnalyticsFeature01": "Resolve crashes efficiently with detailed insights into bug locations, users’ interactions and users’ device information.",
+        "CrashAnalyticsFeature03": "Elevate your websiteʼs performance with real-time crash alerts and scheduled reports to stay informed and ready to resolve bugs quickly.",
+        "CustomReportsFeature01": "Create analytics reports customised to your specific business goals and KPIs, ensuring they focus on the most relevant data.",
+        "CustomReportsFeature02": "Drill down into specific data for deeper insights into user behaviour and engagement.",
+        "CustomReportsFeature03": "Save time and resources by automating report generation, enabling real-time monitoring, and providing cost-effective analytics solutions without the need for third-party tools.",
+        "FormAnalyticsFeature01": "Understand how users engage with your forms to find areas of improvement.",
+        "FormAnalyticsFeature02": "Identify common user errors and validation issues within your forms, to enhance user satisfaction and conversion rates.",
+        "FormAnalyticsFeature03": "Gain deeper insights into how different user groups interact with your forms and tailor your strategies accordingly.",
+        "FunnelsFeature01": "Identify and address drop-off points to improve conversion rates.",
+        "FunnelsFeature02": "Analyse user behaviour based on various criteria, and get valuable insights into how different user groups interact with your site.",
+        "FunnelsFeature03": "Make informed decisions by identifying optimisation opportunities, enhancing user engagement, and driving revenue growth through a deeper understanding of user journeys.",
+        "HeatmapsFeature01": "Get visual representations of user interactions on your website, making it easy to understand how visitors engage with your content.",
+        "HeatmapsFeature02": "Get actionable data to optimise your website's layout, content placement, and user experience.",
+        "HeatmapsFeature03": "Identify and address user behaviour patterns, to increase conversion rates and achieve better results from your digital efforts.",
+        "MediaAnalyticsFeature01": "Get detailed insights into user engagement with audio and video content, to help you understand what resonates with your audience.",
+        "MediaAnalyticsFeature02": "Start fine-tuning your content strategy right away, no complex configurations required.",
+        "MediaAnalyticsFeature03": "See who, how much, and which parts of your media visitors have consumed and which content contributes the most value to your business.",
+        "SessionRecordingsFeature01": "See how visitors interact with your site in real-time and uncover valuable insights to improve user experience.",
+        "SessionRecordingsFeature02": "Identify barriers and successful user journeys, leading to higher conversion rates.",
+        "SessionRecordingsFeature03": "Gain in-depth insights into how users engage with specific content, forms, or elements, to tailor your content and design to better meet user preferences."
```
The verbiage seems a bit inconsistent here with both `visitors` and `users`  being used interchangeably when these terms already mean different things in Matomo. For example Funnels, 'Analyse user behaviour...' could cause confusion as to whether it only applies to identified users or can it also analyze visitor behaviour?

michalkleiner replied:
The copy is taken from the designs done by product and marketing, so I didn't question that.

@Stan-vw any tweaks to the content to tweak the terminology?

Stan-vw replied:
Good catch, I agree that we should stick with `visitor`. As far as I identified we can replace every mention of `user` by `visitor` in these lines.

bx80 reviewed (changes requested):
The tests are all passing now :+1: 
I'm not seeing the promo menu items appearing on the left menu though.

michalkleiner committed:
Use full plugin names in category ids

michalkleiner committed:
Disable Professional Services ads for UI tests apart from the plugin itself

michalkleiner committed:
Fix typo in test method name

michalkleiner committed:
Add new widget categories to widgets test

michalkleiner committed:
Add integration test without professional services widgets/categories

michalkleiner committed:
Update expected list of report pages returned by the API in a system test

Commit:
Added missing externalLink method reference to the vue global properties

bx80 committed:
built vue files

michalkleiner committed:
Use test.vars. approach to disable professional services for UI tests as needed

This is needed as config override from UI tests has lower priority than ui-test.php file overrides.

michalkleiner committed:
Remove test dependency on Advertising class

It seems the DI config isn't passed to the class properly during the UI tests
so using its static method using config provided by the DI works better.

michalkleiner committed:
Merge branch 'm21432-externalrawlink-vue-fix' into dev-17228-premium-plugins-promos-5x

michalkleiner commented:
Already including latest `5.x-dev` with the fix from #21435 to make it easier for the merge when #21435 is merged first.

michalkleiner committed:
Reverse logic so that Professional Services are disabled for UI tests by default and only enabled when required

michalkleiner committed:
Ensure page is reloaded after config change

michalkleiner committed:
Merge branch '5.x-dev' into dev-17228-premium-plugins-promos-5x

michalkleiner committed:
Update screenshot from CI run

michalkleiner committed:
Reset UI tests config before each test

Tidying up the config within the test where it's set didn't work.

michalkleiner committed:
Move mouse out of viewport to prevent any unwanted hover effects

michalkleiner committed:
Update tests from CI

michalkleiner committed:
Update language to use visitor instead of user

michalkleiner committed:
Update UI test screenshots after content change

michalkleiner commented:
@matomo-org/core-reviewers this is finally good for another review.

michalkleiner commented on tests/UI/expected-screenshots/Menus_mainmenu_lower_clicked.png:
Updating this test as there was a hover over the menu visible while it probably shouldn't have been (and moving the mouse out of viewport in the `openMenuItem` changed that) to be in line with other UI tests in this suite.
